lib/EbookDownloadSummary.php

```diff
@@ -0,0 +1,52 @@
+<?
+
+use Safe\DateTimeImmutable;
+
+class EbookDownloadSummary{
+	public int $EbookId;
+	public DateTimeImmutable $Date;
+	public int $DownloadCount = 0;
+	public int $BotDownloadCount = 0;
+
+	public function __construct(int $ebookId, DateTimeImmutable $date){
+		$this->EbookId = $ebookId;
+		$this->Date = $date;
+	}
+
+	/**
+	 * @throws Exceptions\InvalidEbookDownloadSummaryException
+	 */
+	public function Validate(): void{
+		$error = new Exceptions\InvalidEbookDownloadSummaryException();
+
+		if($this->DownloadCount < 0){
+			$error->Add(new Exceptions\InvalidEbookDownloadCountException('Invalid EbookDownloadSummary DownloadCount: ' . $this->DownloadCount));
+		}
+
+		if($this->BotDownloadCount < 0){
+			$error->Add(new Exceptions\InvalidEbookDownloadCountException('Invalid EbookDownloadSummary BotDownloadCount: ' . $this->BotDownloadCount));
+		}
+
+		if($error->HasExceptions){
+			throw $error;
+		}
+	}
+
+	/**
+	 * @throws Exceptions\InvalidEbookDownloadSummaryException
+	 */
+	public function Create(): void{
+		$this->Validate();
+
+		Db::Query('
+			INSERT into EbookDownloadSummaries (EbookId, Date, DownloadCount, BotDownloadCount)
+			values (?,
+				?,
+				?,
+				?)
+			on duplicate key update
+				DownloadCount = values(DownloadCount),
```

Comment:
Should we use `value()` (singular) instead of `values()`? [See here](https://mariadb.com/kb/en/values-value/). It looks like `values()` is backwards compatible for now, but it might not be in the future, so it might be better to change to singular.

Reply:
Done.